pkg/proc/gdbserial/undo.go

```diff
@@ -2,19 +2,318 @@ package gdbserial
 
 import (
 	"bytes"
+	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"os/exec"
+	"os/user"
 	"path/filepath"
 	"runtime"
+	"sort"
 	"strconv"
 	"strings"
 	"syscall"
 
 	"github.com/undoio/delve/pkg/proc"
 )
 
+// State relating to an Undo "session" - this is runtime state that corresponds with.
+type undoSession struct {
+	checkpointLastId int
+	checkpoints      map[int]proc.Checkpoint
+}
+
+// Create a new undoSession structure, which captures the runtime state relating to an on-disk
+// "session" file - currently only bookmarks are supported.
+func newUndoSession() *undoSession {
+	return &undoSession{
+		checkpointLastId: 1,
+		checkpoints:      make(map[int]proc.Checkpoint),
+	}
+}
+
+// Validate a checkpoint note to ensure easy interopability with UDB bookmarks.
+func validateCheckpointNote(where string) error {
+	if where == "" {
+		return errors.New("Checkpoint note cannot be empty.")
+	}
+
+	// Perform matching checking to UDB's bookmark creation code, to ensure Delve notes
+	// translate properly into UDB bookmarks when we save the session.
+	firstChar := where[0:1]
+	if firstChar == " " {
+		return errors.New("Checkpoint note cannot start with a space.")
+	} else if _, err := strconv.Atoi(firstChar); err == nil {
+		return errors.New("Checkpoint note cannot start with a digit.")
+	} else if firstChar == "," || firstChar == "-" || firstChar == "$" {
+		return fmt.Errorf("Checkpoint note cannot start with character: %c", where[0])
+	} else {
+		firstWord := strings.Split(where, " ")[0]
+		reserved := []string{"time", "undo", "redo", "inferior", "start", "annotation",
+			"end", "bookmark", "event", "pc", "wallclock"}
```

Comment:
I suggest keeping lists like this in alphabetical order (unless there is a reason to use some other order) as this makes it easier to check for missing items or duplicates.